expression/builtin_cast.go

```diff
@@ -522,11 +522,18 @@ type builtinCastIntAsDurationSig struct {
 }
 
 func (b *builtinCastIntAsDurationSig) evalDuration(row []types.Datum) (res types.Duration, isNull bool, err error) {
-	val, isNull, err := b.args[0].EvalInt(row, b.getCtx().GetSessionVars().StmtCtx)
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	val, isNull, err := b.args[0].EvalInt(row, sc)
 	if isNull || err != nil {
 		return res, isNull, errors.Trace(err)
 	}
 	res, err = types.ParseDuration(strconv.FormatInt(val, 10), b.tp.Decimal)
+	if types.ErrTruncatedWrongVal.Equal(err) {
+		err = sc.HandleTruncate(err)
+		if res == types.ZeroDuration {
```

Comment:
```go
if err != nil {
    return res, true, errors.Trace(err)
}
```

1. move `if res == types.ZeroDuration {` out of `if types.ErrTruncatedWrongVal.Equal(err) {` statement
2. if res is a zero duration, should we throw a zero duration error ?

Reply:
1. zero duration is valid value. so I need to check ErrTruncatedWrongVal err to distinguish invalid zero from valid zero duration.
```
mysql> select cast('000000' as time) ;
+------------------------+
| cast('000000' as time) |
+------------------------+
| 00:00:00               |
+------------------------+
1 row in set (0.00 sec)
```

2. expression.handleInvalidTimeError appended warning message, but was not compatible with mysql's behavior. ErrInvalidTimeFormat does not contains any params.
mysql:
```
mysql> select cast('123080' as time) ;
+------------------------+
| cast('123080' as time) |
+------------------------+
| NULL                   |
+------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+------------------------------------------+
| Level   | Code | Message                                  |
+---------+------+------------------------------------------+
| Warning | 1292 | Truncated incorrect time value: '123080' |
+---------+------+------------------------------------------+
1 row in set (0.00 sec)
```

Reply:
@zz-jason 

1. with ErrTruncatedWrongVal and zero duration, set isNull to true, append warning or not, then will show <nil>.
2. with ErrTruncatedWrongVal and not zero duration, the over range case,  append warning or not.
3. zero duration without error, the normal case, just pass.